tests/Tests/Directory.hs

```diff
@@ -0,0 +1,66 @@
+module Tests.Directory where
+
+import Test.Framework
+import Test.Framework.Providers.HUnit
+import Test.HUnit.Base
+import Control.Monad (filterM, zipWithM)
+import Data.Function (on)
+import Data.List ((\\), sort)
+import qualified Data.DList as DList
+import System.Directory (doesDirectoryExist)
+import System.FilePath (takeBaseName)
+
+import System.FilePath.Glob.Base
+import System.FilePath.Glob.Directory
+import System.FilePath.Glob.Utils
+
+tests = testGroup "Directory"
+   [ testCase "includeUnmatched" caseIncludeUnmatched
+   , testCase "onlyMatched" caseOnlyMatched
+   ]
+
+caseIncludeUnmatched = do
+   let pats = ["**/D*.hs", "**/[MU]*.hs"]
+   everything <- (DList.toList <$> getRecursiveContents "System")
+                 >>= removeDirectories
+   let expectedMatches =
+          [ [ "System/FilePath/Glob/Directory.hs" ]
+          , [ "System/FilePath/Glob/Match.hs"
+            , "System/FilePath/Glob/Utils.hs"
+            ]
+          ]
+   let everythingElse = everything \\ concat expectedMatches
+
+   result <- globDirWith (GlobOptions matchDefault True)
+                         (map compile pats)
+                         "System"
+   zipWithM assertEqualUnordered expectedMatches (fst result)
+
+   case snd result of
+       Nothing ->
+          assertFailure "Expected Just a list of unmatched files"
+       Just unmatched -> do
+          noDirs <- removeDirectories unmatched
+          assertEqualUnordered everythingElse noDirs
+
+caseOnlyMatched = do
+   let pats = ["**/D*.hs", "**/[MU]*.hs"]
+   let expectedMatches =
+          [ [ "System/FilePath/Glob/Directory.hs" ]
+          , [ "System/FilePath/Glob/Match.hs"
+            , "System/FilePath/Glob/Utils.hs"
+            ]
+          ]
+
+   result <- globDirWith globDefault
+                         (map compile pats)
+                         "System"
+
+   zipWithM assertEqualUnordered expectedMatches (fst result)
+   assertEqual "" Nothing (snd result)
+
+assertEqualUnordered :: (Ord a, Show a) => [a] -> [a] -> Assertion
+assertEqualUnordered = assertEqual "" `on` sort
+
+removeDirectories :: [FilePath] -> IO [FilePath]
+removeDirectories = filterM (fmap not . doesDirectoryExist)
```

Comment:
I don't think you should need to do this: rather remove the only the root directory, i.e. `"System"`.

I took a look and doing this breaks the last assertion in `caseIncludeUnmatched` because `unmatched` starts with `"System/FilePath/"` instead of `"System/FilePath"`. This is actually an unrelated bug (your testcases were immediately useful, thanks!): for example, changing the first pattern to `**///D*.hs` gives:

```
expected: ["System/FilePath/Glob/Directory.hs"]
 but got: ["System/FilePath///Glob/Directory.hs"]
```

What we actually should be getting is `["System/FilePath/Glob///Directory.hs"]` (as `zsh` does) — the slashes are replicated in the wrong place. The incorrect slashing is in `matchTypedAndGo`, on line 205 here:

```
               True | isDir  -> do
                  contents <- getRecursiveContents
                                 (absPath ++ replicate n pathSeparator)
```

Replicating the path separators here is incorrect: it should happen at the deepest level of `getRecursiveContents`, not the first. I'm not sure if that can be done in a non-awkward fashion, though...

You don't have to fix this issue if you don't want, but I'd rather see you only remove `"System"`, and if we're sticking with this slashes bug then do some sort of hack and e.g. leave a FIXME comment about the `"System/FilePath"` vs. `"System/FilePath/"` issue.

Reply:
Ok, sounds good - I think I will go for the hack for now, to avoid this PR getting too complex, and maybe try to address the slashes bug later.